cpp/src/strings/regex/regex.inl

```diff
@@ -40,95 +33,101 @@ namespace detail {
  * reflected here. The regexec function updates and manages this state data.
  */
 struct alignas(8) relist {
-  int16_t size{};
-  int16_t listsize{};
-  int32_t reserved;
-  int2* ranges{};       // pair per instruction
-  int16_t* inst_ids{};  // one per instruction
-  u_char* mask{};       // bit per instruction
-
-  CUDF_HOST_DEVICE inline static int32_t data_size_for(int32_t insts)
+  /**
+   * @brief Compute the memory size for the state data.
+   */
+  constexpr inline static std::size_t data_size_for(int32_t insts)
   {
     return ((sizeof(ranges[0]) + sizeof(inst_ids[0])) * insts) + ((insts + 7) / 8);
   }
 
-  CUDF_HOST_DEVICE inline static int32_t alloc_size(int32_t insts)
+  /**
+   * @brief Compute the aligned memory allocation size.
+   */
+  constexpr inline static std::size_t alloc_size(int32_t insts)
   {
-    int32_t size = sizeof(relist);
-    size += data_size_for(insts);
-    size = ((size + 7) / 8) * 8;  // align it too
-    return size;
+    return cudf::util::round_up_unsafe<size_t>(data_size_for(insts) + sizeof(relist),
+                                               sizeof(ranges[0]));
   }
 
-  CUDF_HOST_DEVICE inline relist() {}
+  struct alignas(16) restate {
+    int2 range;
+    int32_t inst_id;
+    int32_t reserved;
+  };
 
-  CUDF_HOST_DEVICE inline relist(int16_t insts, u_char* data = nullptr) : listsize(insts)
+  __device__ __forceinline__ relist(int16_t insts, u_char* data = nullptr)
+    : masksize((insts + 7) / 8)
   {
     auto ptr = data == nullptr ? reinterpret_cast<u_char*>(this) + sizeof(relist) : data;
     ranges   = reinterpret_cast<int2*>(ptr);
-    ptr += listsize * sizeof(ranges[0]);
+    ptr += insts * sizeof(ranges[0]);
     inst_ids = reinterpret_cast<int16_t*>(ptr);
-    ptr += listsize * sizeof(inst_ids[0]);
+    ptr += insts * sizeof(inst_ids[0]);
     mask = ptr;
     reset();
   }
 
-  CUDF_HOST_DEVICE inline void reset()
+  __device__ __forceinline__ void reset()
   {
-    memset(mask, 0, (listsize + 7) / 8);
+    memset(mask, 0, masksize);
```

Comment:
Possible future optimization: This is almost certainly doing 1B stores. Using a wider 4B/8B/16B writes could possibly help. Do we know what the alignment of the `mask` pointer will be? Can we pad it such that each thread's mask pointer will have a higher alignment than 1B?

Reply:
I feel the overhead here may not be worth it. There is a bit per instruction so even a very large pattern of say 90 instructions would only be 12 bytes.